### PR TITLE
Automate playground compiler

### PR DESCRIPTION
**Pull request by aspeddro**
~~I'm going to split this PR into two parts because the second requires NextJS version 12. Waiting for #645 to finish the second part~~.

Part 2 #565:

- [x] Introduce a protected revalidation route, as stated in the NextJS docs: https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#using-on-demand-revalidation
- [x] Introduce a secret for the revalidation logic and store it in the rescript-lang.org secrets https://github.com/rescript-lang/rescript-compiler/pull/6449

### CODE REVIEW

**Comment by aspeddro** [comment]
This can be merged, part two is independent.

File: pages/try.js [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
-const Try = dynamic(() => import("src/Try.mjs"), {
+export { getStaticProps } from "src/Try.mjs";
+import Try from "src/Try.mjs";
+
+const Playground = dynamic(() => import("src/Playground.mjs"), {
   ssr: false,
-  //loading: () => <div> Loading... </div>
-});
+  loading: () => <span>Loading...</span>,
+})
```
Comment (zth) [inline pages/try.js:9]: Not for now, but this should be possible to move into ReScript now that we have dynamic imports, right?

Comment (aspeddro) [reply pages/try.js:9]: yup

**Comment by aspeddro** [comment]
rescript-compiler repo PR https://github.com/rescript-lang/rescript-compiler/pull/6449

Now we need to pass the environment variable `NEXT_REVALIDATE_SECRET_TOKEN` in the vercel deploy

File: src/others/Revalidate.res [hunk 2]
```diff
@@ -0,0 +1,34 @@
+module Req = {
+  type query = {secret: string}
+
+  type req = {query: query}
```
Comment (aspeddro) [inline src/others/Revalidate.res:4]: query is `Js.Dict.t`

File: src/others/Revalidate.resi [hunk 3]
```diff
@@ -0,0 +1,8 @@
+module Req: {
+  type req
+}
+module Res: {
+  type res
+}
+@val external process: 'a = "process"
```
Comment (aspeddro) [inline src/others/Revalidate.resi:7]: remove

**Comment by aspeddro** [comment]
Ready for review

**Comment by zth** [comment]
@aspeddro I haven't really been in this part of the code yet. Who'd be good to review you think?

**Comment by aspeddro** [comment]
@ryyppy

File: src/common/Util.res [hunk 4]
```diff
@@ -84,3 +84,76 @@ module Date = {
     dateTimeFormat("en-US", {"month": "short", "day": "numeric", "year": "numeric"})->format(date)
   }
 }
+
+/**
+Takes a `version` string starting with a "v" and ending in major.minor.patch or
+major.minor.patch-prerelease.identifier (e.g. "v10.1.0" or "v10.1.0-alpha.2")
+*/
+module Semver = {
```
Comment (ryyppy) [inline src/common/Util.res:92]: Any particular reason why this module doesn't reside in the `CompilerManagerHook` module? Seems like the only module where this is used, and it seems to be quite tailored to the compiler management use-case.

Also, the module description seems to rather describe `Semver.parse`.

**Review by ryyppy — APPROVED** [review]
LGTM. I added some nitpicks, but if you feel this is the way to go, we can also just merge.

I already added the relevant `NEXT_REVALIDATE_SECRET_TOKEN` secret to the Vercel settings and the rescript-compiler repo secrets.

So as soon as we merge this, everything should just work, I assume.

File: src/common/Util.res (hunk 4 above)
Comment (aspeddro) [reply src/common/Util.res]: Moved to CompilerManagerHook

**Comment by aspeddro** [comment]
@ryyppy the deployment failed. I can't see the log either

image

**Comment by ryyppy** [comment]
had to approve the deployment

**Review by cknitt — APPROVED** [review]

**Pull request merged**